### ISSUE TITLE
About FVD usage. How did conditional frames added into the calculation of FVD? 

### ISSUE THREAD

**Issue by JunyaoHu** [voletiv/mcvd-pytorch#19]
# Question

Hi, I come back again haha... Our team want to follow your nice work so now I want to figure out your FVD metrics setting.

For example, I use your pre-trained KTH model to do video prediction task, and use the setting of `kth64_big.yaml` to predict videos. As the config says:

https://github.com/voletiv/mcvd-pytorch/blob/451da2eb635bad50da6a7c03b443a34c6eb08b3a/configs/kth64_big.yml#L65-L70

https://github.com/voletiv/mcvd-pytorch/blob/451da2eb635bad50da6a7c03b443a34c6eb08b3a/configs/kth64_big.yml#L32

It means model should use 10 conditional frames (past frames) to generate 5 predicted frames (future frames). **If we want to prediction 20 frames, when we calculate FVD, do we calculate on 30 frames with 10 conditional frames and 20 predicted frames, or only calculate on 20 predicted frames?** 

When I read other papers, they usually calculate the FVD of predicted frames with conditional frames. **If it is true, what are the considerations behind adding conditional frames in the calculation?**, while SSIM, PSNR, LPIPS do not calculate like this, because they are image metrics not video metircs?

Hope to your early reply! Thank you for your nice contribution again!

# Experiment Results

And I (1) use your code to re-train a new model with 200,000 steps and (2) directedly use your pre-trained model to (3) compare these metrics with your paper. It seems to the metrics FVD is the concat result of conditional frames and predicted frames. 

![image](https://user-images.githubusercontent.com/67564714/226627200-38410fe9-1bf1-4335-82f6-d58cb38e7693.png)

# Code Reading

In validation stage, you sample from model by calling function `video_gen()`

https://github.com/voletiv/mcvd-pytorch/blob/226a3fd1601a6fde1c59d01fb22f82cb37b3b8c4/runners/ncsn_runner.py#L489-L497

**there, `real_fvd` is the concat result of `cond_original` and `real`, and `fake_fvd` also.** then get `real_embeddings` and `fake_embeddings`, we input them to calculate FVD. 

https://github.com/voletiv/mcvd-pytorch/blob/226a3fd1601a6fde1c59d01fb22f82cb37b3b8c4/runners/ncsn_runner.py#L1925-L1953

https://github.com/voletiv/mcvd-pytorch/blob/226a3fd1601a6fde1c59d01fb22f82cb37b3b8c4/runners/ncsn_runner.py#L2234-L2239

# Reference
SLAMP
https://github.com/kaanakan/slamp/blob/4f5fc0707a4843d34dd1cb98f4939f1357e05183/calculate_fvd.py#L25-L31

SRVP
https://github.com/edouardelasalles/srvp/blob/3e90a748db04d182290132163fea5b0410ea2452/test.py#L295-L302

**Comment by JunyaoHu** [comment]
Hello, the above problem has been solved. We will add 10 frames of the original video into the calculation.
But... For example, as for MCVD concat past-future-mask, it have cond=5, future=5, and trained on k=5, how to calculate it if we want to predict 40 frames, it means we should do (40/5)=8 times autoregression also, and then add cond=5 to total videos, then calculate the FVD of 45 frames?

Is that fair compared to the results above which calculate 50=cond10+pred40? (only for prediction task)

![image](https://user-images.githubusercontent.com/67564714/228184743-4cd44b21-38e3-4d3a-8c32-ba6275d88e3d.png)

**Comment by AlexiaJM** [comment]
Hi JunyaoHu, sorry for the delay. 

When you do condition on 10 past frames, but predict the 5 next frames. So it does:
- 1 2 3 4 5 6 7 8 9 10 -> 11 12 13 14 15
- 6 7 8 9 10 11 12 13 14 15 -> 16 17 18 19 20
....
- 36 37 38 39 40 41 42 43 44 45 -> 46 47 48 49 50

When conditioning on 5 to predict 5 it does:
 1 2 3 4 5 -> 6 7 8 9 10
 ....
 36 37 38 39 40 -> 41 42 43 44 45
 
**Both predict 40 frames** which is fair, but the second one is at a disadvantage since it only see 5 frames at a time. Since our model at a disadvantage, its okay since we are not unfair to competitors. 

There is no advantage of going up to 50 frames if you only predict 40 of them.

**Comment by JunyaoHu** [comment]
OK, very thanks！
